agent/system_metrics.go

```diff
@@ -0,0 +1,267 @@
+package agent
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/evergreen-ci/evergreen/model/task"
+	metrics "github.com/evergreen-ci/timber/system_metrics"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/recovery"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+)
+
+// dataFormat describes how the time series data is stored.
+type dataFormat int32
+
+// Valid dataFormat values.
+const (
+	dataFormatText dataFormat = 0
+	dataFormatFTDC dataFormat = 1
+	dataFormatBSON dataFormat = 2
+	dataFormatJSON dataFormat = 3
+	dataFormatCSV  dataFormat = 4
+)
+
+func (f dataFormat) validate() error {
+	switch f {
+	case dataFormatText, dataFormatFTDC, dataFormatBSON, dataFormatJSON, dataFormatCSV:
+		return nil
+	default:
+		return errors.New("invalid schema type specified")
+	}
+}
+
+// metricCollector is an interface representing an object that can collect
+// a single system metric over a series of time steps.
+type metricCollector interface {
+	// Name returns a string indicating the type of metric collected, such as "uptime".
+	Name() string
+	// Format returns the format of the collected data.
+	Format() dataFormat
+	// Collect returns the value of the collected metric when the function is called.
+	Collect() ([]byte, error)
+}
+
+// systemMetricsCollector handles collecting an arbitrary set of system
+// metrics at a fixed interval and streaming them to cedar.
+type systemMetricsCollector struct {
+	mu              sync.Mutex
+	stream          sync.WaitGroup
+	close           sync.WaitGroup
+	streamingCancel context.CancelFunc
+	taskOpts        *metrics.SystemMetricsOptions
+	streamOpts      *metrics.StreamOpts
+	interval        time.Duration
+	collectors      []metricCollector
+	client          *metrics.SystemMetricsClient
+	catcher         grip.Catcher
+	id              string
+	closed          bool
+}
+
+// systemMetricsCollectorOptions are the required values for creating a new
+// systemMetricsCollector.
+type systemMetricsCollectorOptions struct {
+	Task       *task.Task
```

Comment:
Small nit: I don't think these have to be exported since it is a private struct used only within this package.